src/puppetlabs/puppetdb/pql/transform.clj

```diff
@@ -0,0 +1,123 @@
+(ns puppetlabs.puppetdb.pql.transform
+  (:require [clojure.string :as str]
+            [puppetlabs.puppetdb.cheshire :as json]))
+
+(defn transform-from
+  ([entity]
+   ["from" entity])
+  ([entity arg2]
+   ["from" entity arg2])
+  ([entity extract expression]
+   ["from" entity (apply vector (concat extract [expression]))])
+  ([entity extract expression modifier]
+   ["from" entity (apply vector (concat extract [expression modifier]))]))
+
+(defn transform-subquery
+  ([entity]
+   ["subquery" entity])
+  ([entity arg2]
+   ["subquery" entity arg2]))
+
+(defn transform-extract
+  [& args]
+  ["extract" (apply vector args)])
+
+(defn transform-expr4
+  ;; Single arg? collapse
+  ([data] data)
+  ;; Multiple args? turn it into an or statement
+  ([data & args] (vec (concat ["or" data] args))))
+
+(defn transform-expr3
+  ;; Single arg? collapse
+  ([data] data)
+  ;; Multiple args? turn it into an and statement
+  ([data & args] (vec (concat ["and" data] args))))
+
+(defn transform-expr2
+  ;; Single arg? Just collapse the :expr2 and pass back the data,
+  ;; closing the nesting.
+  ([data] data)
+  ;; Two args? This means :not [data] so convert it into a "not"
+  ([_ data] ["not" data]))
+
+(defn transform-function
+  [entity args]
+  (concat ["function" entity] args))
```

Comment:
On that note a `(def concatv (comp vec concat))` would make a good helper function for this file.